src/include/daos/tls.h

```diff
@@ -0,0 +1,124 @@
+ * - a server-side API (see header files suffixed by "_srv") used for
+ *   inter-module direct calls.
+ *
+ * For now, all loaded modules are assumed to be trustful, but sandboxes can be
+ * implemented in the future.
+ */
+/*
+ * Thead-local storage
+ */
+struct daos_thread_local_storage {
+	uint32_t	dtls_tag;
+	void		**dtls_values;
+};
+
+enum daos_module_tag {
+	DAOS_SYS_TAG    = 1 << 0, /** only run on system xstream */
+	DAOS_TGT_TAG    = 1 << 1, /** only run on target xstream */
+	DAOS_RDB_TAG    = 1 << 2, /** only run on rdb xstream */
+	DAOS_OFF_TAG    = 1 << 3, /** only run on offload/helper xstream */
+	DAOS_CLI_TAG    = 1 << 4, /** only run on client stack */
+	DAOS_SERVER_TAG = 0xff,   /** run on all xstream */
+};
+
+/* The module key descriptor for each xstream */
+struct daos_module_key {
+	/* Indicate where the keys should be instantiated */
+	enum daos_module_tag dmk_tags;
+
+	/* The position inside the daos_module_keys */
+	int dmk_index;
+	/* init keys for context */
+	void *(*dmk_init)(int tags, int xs_id, int tgt_id);
+
+	/* fini keys for context */
+	void (*dmk_fini)(int tags, void *data);
+};
+
+#define DAOS_MODULE_KEYS_NR 10
+extern pthread_key_t dc_tls_key;	/* client side TLS key */
+extern pthread_key_t dss_tls_key;	/* server side TLS key */
+extern struct daos_module_key *daos_module_keys[DAOS_MODULE_KEYS_NR];
+
+static inline struct daos_thread_local_storage *
+dss_tls_get()
+{
+	return (struct daos_thread_local_storage *)
+		pthread_getspecific(dss_tls_key);
+}
+
+static inline struct daos_thread_local_storage *
+dc_tls_get()
+{
+	return (struct daos_thread_local_storage *)
+		pthread_getspecific(dc_tls_key);
+}
+
+/* For now TLS is only enabled if metrics are enabled */
+#define METRIC_DUMP_ENV "DAOS_METRIC_DUMP_ENV"
+#define DAOS_CLIENT_METRICS_ENV	"DAOS_CLIENT_METRICS"
+#define DAOS_CLIENT_METRICS_RETAIN_ENV	"DAOS_CLIENT_METRICS_RETAIN"
+extern bool client_metric;
+extern bool client_metric_retain;
+
+/**
+ * Get value from context by the key
+ *
+ * Get value inside dtls by key. So each module will use this API to
+ * retrieve their own value in the thread context.
+ *
+ * \param[in] dtls	the thread context.
+ * \param[in] key	key used to retrieve the dtls_value.
+ *
+ * \retval		the dtls_value retrieved by key.
+ */
+static inline void *
+daos_module_key_get(struct daos_thread_local_storage *dtls,
+		    struct daos_module_key *key)
+{
+	D_ASSERT(key->dmk_index >= 0);
+	D_ASSERT(key->dmk_index < DAOS_MODULE_KEYS_NR);
+	D_ASSERT(daos_module_keys[key->dmk_index] == key);
+	D_ASSERT(dtls != NULL);
+
+	return dtls->dtls_values[key->dmk_index];
+}
+
+#define dss_module_key_get daos_module_key_get
+#define dss_register_key   daos_register_key
+#define dss_unregister_key daos_unregister_key
+#define dss_module_info    daos_module_info
+#define dss_module_tag    daos_module_tag
+#define dss_module_key    daos_module_key
+#define dss_thread_local_storage daos_thread_local_storage
```

Comment:
Are these appropriate for a public interface or should they be private?

Reply:
This patch move TLS to both client and server side, so some structures are renamed from dss to daos, and these definition is for avoiding changed too much name on each name. Hmm, I am not sure where to put these, if not with the new TLS definition.